lib/pbench/server/api/resources/__init__.py

```diff
@@ -731,8 +743,144 @@ def _check_authorization(self, user: Union[str, None], access: Union[str, None])
                 # READ, the user owns the data, or the user has ADMIN role.
                 pass
 
-    def _get_metadata(
-        self, controller: str, name: str, requested_items: List[str]
+    def _build_sql_query(self, parameters: JSON, base_query: Query) -> Query:
+        """
+        Extend a SQLAlchemy Query with additional terms applying specified
+        owner and access constraints from the parameters JSON.
+
+        NOTE: This method is not responsible for authorization checks; that's
+        done by the `ApiBase._check_authorization()` method. This method is
+        only concerned with generating a query to restrict the results to the
+        set matching authorized user and access values. This method will build
+        queries that do not reflect the input when given some inputs that must
+        be prevented by authorization checks.
+
+        Specifically, when asked for "access": "private" on behalf of an
+        unauthenticated client or on behalf of an authenticated but non-admin
+        client for a different user, the generated query will use "access":
+        "public". These cases are designated below with "NOTE(UNAUTHORIZED)".
+
+        Specific cases for user and access values:
+
+            {}: defaulting both user and access
+                All private + public regardless of owner
+
+                ADMIN: all datasets
+                AUTHENTICATED as drb: owner:drb OR access:public
+                UNAUTHENTICATED (or non-drb): access:public
+
+            {"user": "drb"}: defaulting access
+                All datasets owned by "drb"
+
+                ADMIN, AUTHENTICATED as drb: owner:drb
+                UNAUTHENTICATED (or non-drb): owner:drb AND access:public
+
+            {"user": "drb, "access": "private"}: private drb
+                All datasets owned by "drb" with "private" access
+
+                owner:drb AND access:private
+                NOTE(UNAUTHORIZED): owner:drb and access:public
+
+            {"user": "drb", "access": "public"}: public drb
+                All datasets owned by "drb" with "public" access
+
+                NOTE: unauthenticated users are not allowed to query by
+                username, and shouldn't get here, but the query is
+                technically correct if we allowed it.
+
+                owner:drb AND access:public
+
+            {"access": "private"}: private data
+                All datasets with "private" access regardless of user
+
+                ADMIN: all access:private
+                AUTHENTICATED as "drb": owner:"drb" AND access:private
+                NOTE(UNAUTHORIZED): access:public
```

Comment:
Fair enough.  Documenting is good.

Although, I dispute the "lie" part -- we would be returning the list of datasets that the requestor is authorized to see...which is _none_ in that case...but, with this approach, instead of returning nothing we would be returning a result which does not match the request.  (I'm not seeing how returning an empty result would require extra code:  we've already got "extra" code which changes the access to `"public"`...if we can't come up with something better, then we could change the access to something like `"denied"` which should produce the appropriate result.)